expression/builtin_control.go

```diff
@@ -127,34 +150,80 @@ func InferType4ControlFuncs(ctx sessionctx.Context, funcName string, lexp, rexp
-			lhsFlagLen, rhsFlagLen := 0, 0
-			if !lhsUnsignedFlag {
-				lhsFlagLen = 1
-			}
-			if !rhsUnsignedFlag {
-				rhsFlagLen = 1
-			}
-			lhsFlen := lhs.GetFlen() - lhsFlagLen
-			rhsFlen := rhs.GetFlen() - rhsFlagLen
-			if lhs.GetDecimal() != types.UnspecifiedLength {
-				lhsFlen -= lhs.GetDecimal()
-			}
-			if lhs.GetDecimal() != types.UnspecifiedLength {
-				rhsFlen -= rhs.GetDecimal()
-			}
-			flen := maxlen(lhsFlen, rhsFlen) + resultFieldType.GetDecimal() + 1 // account for -1 len fields
-			resultFieldType.SetFlenUnderLimit(flen)
-		} else if evalType == types.ETString {
-			lhsLen, rhsLen := lhs.GetFlen(), rhs.GetFlen()
-			if lhsLen != types.UnspecifiedLength && rhsLen != types.UnspecifiedLength {
-				resultFieldType.SetFlen(mathutil.Max(lhsLen, rhsLen))
+	case ast.Case:
+		if len(args) == 0 {
+			panic("unexpected length 0 of args for casewhen")
+		}
+		fallthrough
+	default:
+		ec, err := CheckAndDeriveCollationFromExprs(ctx, funcName, evalType, args...)
+		if err != nil {
+			return err
+		}
+		resultFieldType.SetCollate(ec.Collation)
+		resultFieldType.SetCharset(ec.Charset)
+		resultFieldType.SetFlag(0)
+		for i := range args {
+			if !types.IsNonBinaryStr(args[i].GetType()) {
+				resultFieldType.AddFlag(mysql.BinaryFlag)
+				break
 			}
+		}
+		// Set retType to BINARY(0) if all arguments are of type NULL.
+		if resultFieldType.GetType() == mysql.TypeNull {
+			types.SetBinChsClnFlag(resultFieldType)
+		}
+	}
+	return nil
+}
+
+// InferType4ControlFuncs infer result type for builtin IF, IFNULL, NULLIF, CASEWHEN, LEAD and LAG.
+func InferType4ControlFuncs(ctx sessionctx.Context, funcName string, args ...Expression) (*types.FieldType, error) {
+	if len(args) == 0 {
+		panic("unexpected length 0 of args")
+	}
+
+	argsNum := len(args)
+	nullFields := make([]*types.FieldType, 0, argsNum)
+	notNullFields := make([]*types.FieldType, 0, argsNum)
+	for i := range args {
+		if args[i].GetType().GetType() == mysql.TypeNull {
+			nullFields = append(nullFields, args[i].GetType())
 		} else {
-			resultFieldType.SetFlen(maxlen(lhs.GetFlen(), rhs.GetFlen()))
+			notNullFields = append(notNullFields, args[i].GetType())
 		}
 	}
+	resultFieldType := &types.FieldType{}
+	if len(nullFields) == argsNum { // all field is TypeNull
+		*resultFieldType = *nullFields[0]
+		// If any of arg is NULL, result type need unset NotNullFlag.
+		tempFlag := resultFieldType.GetFlag()
+		types.SetTypeFlag(&tempFlag, mysql.NotNullFlag, false)
+		resultFieldType.SetFlag(tempFlag)
+		// If both arguments are NULL, make resulting type BINARY(0).
+		resultFieldType.SetType(mysql.TypeString)
+		resultFieldType.SetFlen(0)
+		resultFieldType.SetDecimal(0)
+		types.SetBinChsClnFlag(resultFieldType)
+	} else if len(nullFields) > 0 { // one of field is TypeNull
+		// If any of arg is NULL, result type need unset NotNullFlag.
+		*resultFieldType = *notNullFields[0]
```

Comment:
Whyt the resultFieldType is not `types.AggFieldType` of all input types?

Reply:
good catch!

Reply:
done